devdoc/Metrics.rst

```diff
@@ -0,0 +1,212 @@
+=================
+Metrics Framework
+=================
+
+************
+Introduction
+************
+
+The metrics framework provides a set of classes and templates for implementing a metrics collection strategy
+in the platform.
+
+The framework consists of three main areas. Each is responsible for a portion of the collection and reporting
+operations. These areas are
+
+* Metrics
+* Sinks
+* The Metrics Reporter
+
+*******
+Metrics
+*******
+
+Metrics represent the base value that a component must track. The framework defines a number of metric
+classes for use by components. Each metric class encapsulates all of the necessary logic to handle the
+following:
+
+* Updating the value being tracked
+* Synchronization between updating and collecting
+* Providing measurements to the framework
+
+Components are only responsible for creating metric objects, instrumenting the component to update
+each metric, and adding each metric to the *Metrics Reporter*. The framework handles all remaining
+tasks related to measurement collection and reporting. Components are independent of the collection
+system(s) in use.
+
+Please see the section on *Instrumenting a Component* for more information on component requirements.
+
+Metrics and Measurements
+------------------------
+
+There is an important distinction between a metric and the measurements it provides. A *metric* is
+the low level value defined by the component that tracks some state in the component. An example is
+the number of requests processed. A *measurement* represents a view of the metric.
+A *metric* may provide multiple measurements. For example, a component defines a metric to count the
+number of times some event occurs. The metric, however, provides different measurements of the
+event. For example it may provide the raw count, the number of events since the last time it was
+collected, and the rate of events per second since the last read. The purpose for providing
+different measurements from the base raw value is to support native value types defined by different collection
+systems. Each metric defines its supported measurements.
+
+
+Available Metrics
+-----------------
+
+The framework provides the following metrics.
+
+* Counter
+* Gauge
+
+Counter Metric
+==============
+
+The *Counter* metric defines a monotonically increasing value used to count the occurrences of a component
+defined event. The metric defines the following available measurements:
+* count - The current metric value (default measurement)
+* resetting_count - The count since the last collection
```

Comment:
Also means that you can make the code threadsafe (for async calls to the collectors) with a critsec in the collector rather than needing one in every metric object

Reply:
Thread safety is handled in several ways. 
* Each metric handles synchronization of access to its internal metric state. This allows the component to implement updating the metric state w/o worry of when or how often a collection is done. The current implementations use atomics.
* The reporter uses a mutex to serialize reporting so that only a single sink is executing a collection at a time
* The reporter uses a mutex to control modifications to the metric vector containing of all registered metrics
* Weak pointers are employed to synchronize locking of metric objects to ensure presence before access. Absence triggers removal. This also handles cases where the component may not gracefully fail a process and cleanup any process related metrics.

The above also keeps the sink free from implementing any framework threading dependencies. It only has to worry about sink specific synchronization dependencies.